sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentTests.java

```diff
@@ -115,26 +114,69 @@ public CompletableFuture<TestAsyncResult> runAsync(Supplier<CompletableFuture<Ma
             }
             var mockEngine = (MockEngine) engine;
             var mockMonitor = (MockMonitor) monitor;
-            return this.runner.runAsyncFuture(callback)
+            var helper = new OutputsCaptureHelper(callback);
+            return this.runner.runAsyncFuture(helper::run)
                     .thenApply(ignore -> new TestAsyncResult(
                             ImmutableList.copyOf(mockMonitor.resources),
                             mockEngine.getErrors().stream()
                                     .map(RunException::new)
-                                    .collect(toImmutableList())
-                    ));
+                                    .collect(toImmutableList()),
+                            helper.getOutputs()));
+        }
+
+        public CompletableFuture<TestAsyncResult> tryTestAsync(Consumer<Context> stackFactory) {
+            return this.runAsync(() -> {
+                var context = new TestContext();
+                stackFactory.accept(context);
+                return CompletableFuture.completedFuture(context.getStackOutputs());
+            });
+        }
+
+        public CompletableFuture<ImmutableList<Resource>> testAsync(Consumer<Context> stackFactory) {
+            return this.tryTestAsync(stackFactory).thenApply(result -> {
+                result.throwOnError();
+                return result.resources;
+            });
         }
 
         public static class TestAsyncResult {
             public final ImmutableList<Resource> resources;
             public final ImmutableList<Exception> exceptions;
+            public final Map<String, Output<?>> stackOutputs;
 
-            public TestAsyncResult(ImmutableList<Resource> resources, ImmutableList<Exception> exceptions) {
+            public TestAsyncResult(ImmutableList<Resource> resources,
+                                   ImmutableList<Exception> exceptions,
+                                   Map<String, Output<?>> stackOutputs) {
                 this.resources = resources;
                 this.exceptions = exceptions;
+                this.stackOutputs = stackOutputs;
+            }
+
+            public <T> Output<T> getStackOutput(String name) {
+                if (!this.stackOutputs.containsKey(name)) {
+                    return Output.of(CompletableFuture.failedFuture(new StackOutputNotFoundException(name)));
```

Comment:
this needs to be short lived, until the next refactor, looks like accidental complexity

Reply:
I actually think something like this to be long-lived but perhaps renamed.  We need a clean design for mock testing API, and in this API the user should be able to easily pull out the stack exports to make assertions against. How would it look like if not using this API?

Reply:
This is clean, it's what I've done in my rerfactoring:
```
Stack.outputs()
Stack.output(name, type)
```

Reply:
Not sure about the user, but useful internally for sure.